### ISSUE TITLE
Tracking Issue for llvm-tools-preview

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#85658]
This is a tracking issue for the llvm-tools-preview.

llvm-tools-preview is a rustup component that contains some LLVM tools. At the time of this writing, it contains the following:

* llc
* llvm-ar
* llvm-as
* llvm-cov
* llvm-dis
* llvm-nm
* llvm-objcopy
* llvm-objdump
* llvm-profdata
* llvm-readobj
* llvm-size
* llvm-strip
* opt

### Unresolved Questions

* What will it take to transition out of `-preview`? 

### Implementation history

* #49584 Original issue
* #50336 Added tools
* #51922 Renamed to `-preview`
* #51962 Added llvm-strip
* #53740 Added llvm-readobj
* #59351 Added llvm-ar
* #72000 Moved libLLVM.so around
* #77868 Added llvm-dis, llc, opt
* #78947 Added llvm-cov
* #78968 Added llvm-as

**Comment by lf-** [comment]
Related issue: https://github.com/rust-lang/rust/issues/56371

The use case that I want clang for is the integrated assembler, for assembling .s files in my rust kernel build process. I don't believe that there's currently any way of invoking the llvm based assembler except through the clang driver.

**Comment by hellow554** [comment]
> The use case that I want clang for is the integrated assembler, for assembling .s files in my rust kernel build process.

If you are  using a nightly compiler you could use the [`asm!`](https://doc.rust-lang.org/unstable-book/library-features/asm.html) macro?

**Comment by lf-** [comment]
> > The use case that I want clang for is the integrated assembler, for assembling .s files in my rust kernel build process.
> 
> If you are using a nightly compiler you could use the [`asm!`](https://doc.rust-lang.org/unstable-book/library-features/asm.html) macro?

Not really, if you are a kernel. Generally you want your init code and similarly large pieces of assembly outside of rust files, since you want to create jump tables and such that don't belong to a function, among other things. It's rather hacky to try to do everything with inline assembly.

**Comment by hellow554** [comment]
What about [`global_asm!`](https://doc.rust-lang.org/unstable-book/library-features/global-asm.html) together with `include_str!`? 🤔

**Comment by tv42** [comment]
These seem to be now available as "llvm-tools" (no "-preview"). Did this issue get accidentally left behind?

**Comment by ehuss** [comment]
It was inadvertent that the component without the `-preview` suffix was allowed. This is currently tracked in #119164.

**Comment by ehuss** [comment]
There was a discussion at https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/llvm-tools.20stabilization about the accidental removal of the `-preview` suffix.

I think to summarize: People agree that we should not remove the `llvm-tools` component and break users currently using it. However, it doesn't seem like there is consensus it should be treated as stable. I have posted https://github.com/rust-lang/rustup/pull/3707 to add a comment about the status.

I had some uncertainty about who is actually responsible for this component. I think by default it should probably be the compiler team delegating to wg-llvm if needed.

There is some uncertainty about the directory where the tools currently reside. Currently they are in the target directory, but they are host tools so they should probably be located somewhere else. There should also probably be a `rustc --print` option for locating that directory. (I assume it should *not* be in `$SYSROOT/bin`, since that seems like it could really break things if people have that in their PATH.)

I'll also copy @wesleywiser's last comment, which seemed like a good summary:

> I agree that simply saying "llvm-tools may break at any time" is uncomfortable, but I think the fairest statement is that "llvm-tools are provided as-is from upstream LLVM". In most cases, they aren't going to break all the time or even occasionally. At the same time, we're not going to take on maintaining a tool LLVM has dropped just to fulfil a stability promise. I think the vast majority of users would totally understand this position and not have an issue with that.
> 
> This feels very similar to the Target Tier Policy where we do not promise perpetual support for any particular target, just a level of support in a given release. Breakage in upstream LLVM has even necessitated demoting targets from Tier 2 to Tier 3 because they're simply too broken to be at that level.

**Comment by dancrossnyc** [comment]
I noticed that, on tonight's nightly, the contents of `llvm-tools` appears to be, either in whole or in very large part, in another component: `rustup` complained about conflicts with `llc`. As an experiment, I removed `llvm-tools` and updated and was surprised to see that e.g. `llvm-objdump` was still installed.

Is this expected?

**Comment by ehuss** [comment]
@dancrossnyc, there was a regression in the latest nightly, see https://github.com/rust-lang/rust/issues/130291. It should be fixed in the next nightly tonight.

**Comment by dancrossnyc** [comment]
Good to go. Thank you, Eric!

**Comment by herrwusel** [comment]
I just updated to llvm19 and now grcov is complaining about mismatching versions for llvm-profdata and the raw data. I wondered how the bins are updated and where I could follow updates?

**Comment by yuhui518** [comment]
> I just updated to llvm19 and now grcov is complaining about mismatching versions for llvm-profdata and the raw data. I wondered how the bins are updated and where I could follow updates?

My issue is:
ation profile format version
warning: /tmp/.tmpfHs8e8/cov-21560-15860613078634779016_0_1.profraw: unsupported instrumentation profile format version
error: no profile can be merged
12:42:34 [ERROR] Error while executing llvm tools: Failure while running "/linux/llvm/bin/llvm-profdata" "merge" "-f" "-" "-sparse" "-o" "/tmp/.tmpfHs8e8/0/grcov.profdata"

[root@localhost ut]# cargo -V
warning: `/root/.cargo/config` is deprecated in favor of `config.toml`
note: if you need to support cargo 1.38 or earlier, you can symlink `config` to `config.toml`
cargo 1.81.0 (2dbb1af80 2024-08-20)

[root@localhost bin]#./llvm-profdata show --version
LLVM (http://llvm.org/):
LLVM version 15.0.4
Optimized build.
Default target: x86_64-unknown-linux-gnu


[root@localhost bin]# ./clang --version
clang version 15.0.4 (llvm-project 9176a70a8154b7ab1764e6e94f6ab58bc9f9b189)
Target: x86_64-unknown-linux-gnu

Llvm-profdata and clang is same version，I don't know why this problem occurred？

**Comment by djc** [comment]
I'm guessing it's because your llvm-profdata is out of sync with the LLVM toolchain that Rust used for 1.81, which is much newer. (I don't have 1.81 handy right now, but 1.80 used LLVM 18.1.7; 1.85 uses 19.1.7).
